### PR TITLE
Kubernetes support.

### PR DESCRIPTION
**Pull request by felidadae**

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -0,0 +1 @@
+PYTHONPATH=.
```
Comment (iwankgb) [inline .env]: What is the purpose of this file?

File: owca/kubernetes.py [hunk 2]
```diff
@@ -133,22 +133,32 @@ def get_tasks(self) -> List[Task]:
                 subcgroups_paths=containers_cgroups))
 
         _log_found_tasks(tasks)
+
+        if not _are_all_tasks_of_single_qos(tasks):
+            log.warning("Tasks from namespaces {} are not all in the same qos groups."
```
Comment (iwankgb) [inline owca/kubernetes.py]: We have discussed it today: tasks can be of various QoS class but cgroup structure must be flat.

File: owca/kubernetes.py [hunk 3]
```diff
@@ -133,22 +133,32 @@ def get_tasks(self) -> List[Task]:
...
-        return os.path.join('/kubepods',
-                            '' if qos == 'guaranteed' else qos,
-                            'pod{}'.format(pod_id),
-                            container_id, "")
+        result = os.path.join('/kubepods',
+                              '' if qos == 'guaranteed' else qos,
+                              'pod{}'.format(pod_id),
+                              container_id, "")
+    # Remove last slash from path.
+    if len(result) > 1 and result[-1] == '/':
```
Comment (iwankgb) [inline owca/kubernetes.py]: Why have you decided to change the logic? It was discussed multiple times and such change will probably affect workloads running on Mesos.

File: tests/test_containers.py [hunk 4]
```diff
@@ -230,3 +230,11 @@ def test_containerset_get_pids(*args):
...
     c = container("/t1", [], rdt_enabled=True, resgroup_name='t1')
     assert c.get_allocations() == {}
+
+
+@patch('owca.containers.ResGroup.get_allocations', return_value={})
+@patch('owca.cgroups.Cgroup.get_allocations', return_value={})
+@patch('owca.containers.ContainerSet', spec=ContainerSet)
+def test_containerset_get_allocations(*mock):
+    c = container("/t1", ['/t1/s1', '/t1/s2'], rdt_enabled=True, resgroup_name='t1')
+    assert c.get_allocations() == {}
```
Comment (iwankgb) [inline tests/test_containers.py]: Increase number of tested scenarios, please.

File: owca/kubernetes.py (hunk 3 above)
Comment (felidadae) [reply owca/kubernetes.py]: thats the logic which is contained in mesos module, i know the fact that previously i added the slash on the end can be misleading.

File: .env (hunk 1 above)
Comment (felidadae) [reply .env]: when running pipenv shell this file is automatically sourced

File: owca/kubernetes.py (hunk 2 above)
Comment (felidadae) [reply owca/kubernetes.py]: As we discussed today I will only check if all tasks are in the same qos group while allocations is going be be applied which change cpu.shares.

File: .env (hunk 1 above)
Comment (felidadae) [reply .env]: I will remove this file from this PR, as it has no relation

File: owca/kubernetes.py (hunk 3 above)
Comment (felidadae) [reply owca/kubernetes.py]: https://github.com/ppalucki/owca/blob/allocation-wip-v2/owca/containers.py#L48
here we make assuption than path ends without a slash

File: configs/kubernetes_example_allocator.yaml [hunk 5]
```diff
@@ -0,0 +1,36 @@
+# Copyright (c) 2019 Intel Corporation
```
Comment (ppalucki) [inline configs/kubernetes_example_allocator.yaml]: rebase on allocation-runner and move to proper kubernetes folder

File: owca/runners/allocation.py [hunk 6]
```diff
@@ -112,6 +114,20 @@ def rdt_allocation_value_constructor(rdt_allocation: RDTAllocation,
...
+    # Ignore if not KubernetesNode.
+    if not have_tasks_qos_label(tasks):
+        return
+
+    if not are_all_tasks_of_single_qos(tasks):
+        for task_id, allocation in allocations.items():
+            if AllocationType.SHARES in allocation:
+                raise InvalidAllocations('not all tasks are of the same Kubernetes QOS class '
+                                         'and at least one of the allocation contains '
+                                         'cpu share.')
```
Comment (ppalucki) [inline owca/runners/allocation.py]: Could you add/replace to comment something like this:
- "Mixing QoS classes and shares allocation is not supported! Please reschedule Pod using with the same QoS class for shares allocation."

Comment (felidadae) [reply owca/runners/allocation.py]: Fixed, but I did not add part  `Please reschedule Pod using with the same QoS class for shares allocation.` as imho it is in contradiction with the fact that OWCA is still running.

File: tests/test_containers.py (hunk 4 above)
Comment (felidadae) [reply tests/test_containers.py]: I mock all called functions inside get_allocations. All the logic branches are tested here.

File: owca/kubernetes.py [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+    #   Reference in source code for kubernetes version stable 1.13:
+    #   https://github.com/kubernetes/kubernetes/blob/v1.13.3/pkg/kubelet/cm/cgroup_manager_linux.go#L207
+
+    cgroup_driver: CgroupDriverType = field(
+        default_factory=lambda: CgroupDriverType(CgroupDriverType.CGROUPFS))
+
+    # By default use localhost, however kubelet may not listen on it.
+    kubelet_endpoint: str = 'https://127.0.0.1:10250'
+
+    # Key and certificate to access kubelet API.
```
Comment (ppalucki) [inline owca/kubernetes.py]: is it required or optional ?

File: owca/resctrl_allocations.py [hunk 8]
```diff
@@ -324,6 +325,28 @@ def _parse_schemata_file_row(line: str) -> Dict[str, str]:
     return domains
 
 
+def _compare_rdt_allocation_l3(current: Optional[str], new:Optional[str]):
+    # In case when at least one of the allocations is None return True.
```
Comment (ppalucki) [inline owca/resctrl_allocations.py]: add doc string

Comment (ppalucki) [reply owca/resctrl_allocations.py]: used that for mb as well

File: owca/resctrl_allocations.py [hunk 9]
```diff
@@ -206,9 +206,10 @@ def calculate_changeset(self, current: Optional['RDTAllocationValue']):
             target = current._copy(target_rdt_allocation)
 
             # Prepare changeset
-            # Logic: if new value exists and is different from old one the use new.
+            # Logic: if new value exists and is different from old one use the new.
             if new.rdt_allocation.l3 is not None \
-                    and current.rdt_allocation.l3 != new.rdt_allocation.l3:
+                    and not _compare_rdt_allocation_l3(current.rdt_allocation.l3,
+                                                       new.rdt_allocation.l3):
```
Comment (ppalucki) [inline owca/resctrl_allocations.py]: use compare for mb string as well

File: owca/resctrl_allocations.py [hunk 10]
```diff
@@ -324,6 +325,28 @@ def _parse_schemata_file_row(line: str) -> Dict[str, str]:
...
+        return first.lstrip('0') == second.lstrip('0')
+
+    # Leading underscore indicates that the data is parsed.
+    current_ = _parse_schemata_file_row(current)
+    new_ = _parse_schemata_file_row(new)
+
+    assert set(new_.keys()) & set(current_.keys()) == set(new_.keys()), \
+           "Assume that new allocations domain is subset of current allocations domain."
+
+    for domain in set(current_.keys()) & set(new_.keys()):
```
Comment (ppalucki) [inline owca/resctrl_allocations.py]: do it once,

File: owca/resctrl_allocations.py [hunk 11]
```diff
@@ -324,6 +325,28 @@ def _parse_schemata_file_row(line: str) -> Dict[str, str]:
...
+    # In case when at least one of the allocations is None return True.
+    if current is None or new is None:
+        assert (current is None and new is None) or (new is None)
+        return True
+
+    def _compare_ignoring_leading_zeros(first, second):
+        return first.lstrip('0') == second.lstrip('0')
+
+    # Leading underscore indicates that the data is parsed.
+    current_ = _parse_schemata_file_row(current)
```
Comment (ppalucki) [inline owca/resctrl_allocations.py]: current_domains

File: owca/resctrl_allocations.py [hunk 12]
```diff
@@ -324,6 +325,28 @@ def _parse_schemata_file_row(line: str) -> Dict[str, str]:
     return domains
 
 
+def _compare_rdt_allocation_l3(current: Optional[str], new:Optional[str]):
```
Comment (ppalucki) [inline owca/resctrl_allocations.py]: _is_rdt_allocation_equal

File: owca/resctrl_allocations.py [hunk 13]
```diff
@@ -324,6 +325,28 @@ def _parse_schemata_file_row(line: str) -> Dict[str, str]:
...
+    # In case when at least one of the allocations is None return True.
+    if current is None or new is None:
+        assert (current is None and new is None) or (new is None)
+        return True
+
+    def _compare_ignoring_leading_zeros(first, second):
```
Comment (ppalucki) [inline owca/resctrl_allocations.py]: _is_equal

File: owca/testing.py [hunk 14]
```diff
@@ -181,14 +201,17 @@ def run(self):
     ))
 
 
-def redis_task_with_default_labels(task_id):
+def redis_task_with_default_labels(task_id, subcgroups_paths=[]):
```
Comment (ppalucki) [inline owca/testing.py]: use immultable, and then `or []`

File: owca/kubernetes.py [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+    def has_value(cls, value):
+        return any(value == item.value for item in cls)
+
+
+@dataclass
+class KubernetesNode(Node):
+    # We need to know what cgroup driver is used to properly build cgroup paths for pods.
+    #   Reference in source code for kubernetes version stable 1.13:
+    #   https://github.com/kubernetes/kubernetes/blob/v1.13.3/pkg/kubelet/cm/cgroup_manager_linux.go#L207
+
```
Comment (ppalucki) [inline owca/kubernetes.py]: add full docstring with Args:

Comment (ppalucki) [reply owca/kubernetes.py]: add refernce in /README.rst

File: owca/resctrl_allocations.py (hunk 8 above)
Comment (felidadae) [reply owca/resctrl_allocations.py]: fixed

File: owca/resctrl_allocations.py (hunk 9 above)
Comment (felidadae) [reply owca/resctrl_allocations.py]: fixed

File: owca/resctrl_allocations.py (hunk 10 above)
Comment (felidadae) [reply owca/resctrl_allocations.py]: fixed

File: owca/resctrl_allocations.py (hunk 11 above)
Comment (felidadae) [reply owca/resctrl_allocations.py]: fixed

File: owca/resctrl_allocations.py (hunk 12 above)
Comment (felidadae) [reply owca/resctrl_allocations.py]: _does_reperform_rdt_suballocation fixed

File: owca/resctrl_allocations.py (hunk 13 above)
Comment (felidadae) [reply owca/resctrl_allocations.py]: fixed

File: owca/testing.py (hunk 14 above)
Comment (felidadae) [reply owca/testing.py]: fixed

File: owca/kubernetes.py (hunk 7 above)
Comment (felidadae) [reply owca/kubernetes.py]: fixed, it is optional, and its working when not set

File: owca/kubernetes.py (hunk 15 above)
Comment (felidadae) [reply owca/kubernetes.py]: fixed

**Pull request merged**